Lock Configuration Usage at Construction (#22580)
gradle/gradle: subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

+
image::deprecations.png[Deprecations View of a Gradle Build Scan]
+
This is so you can see any deprecation warnings that apply to your build.
+
Alternatively, you could run `gradle help --warning-mode=all` to see the deprecations in the console, though it may not report as much detailed information.
. Update your plugins.
+
Some plugins will break with this new version of Gradle, for example because they use internal APIs that have been removed or changed. The previous step will help you identify potential problems by issuing deprecation warnings when a plugin does try to use a deprecated part of the API.
+
. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_9.0]]
== Upgrading from 8.x and earlier

=== Warnings that are now errors

=== Removed APIs

=== Potential breaking changes

=== Deprecations

[[custom_configuration_roles]]

> Custom roles have been deprecated. Use pre-defined roles from `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` instead.

Reviewer: I don't think we should be linking to internal classes in the upgrade guide.

Author: I'm going to leave this.  This page is unreachable currently, until we hit Gradle 9.0.  At that point, these classes will have been moved to no longer be internal - so this shouldn't be an issue.

Reviewer: Wouldn't this page become accessible in `8.1`?

Author: Updated.

+
image::deprecations.png[Deprecations View of a Gradle Build Scan]
+
This is so that you can see any deprecation warnings that apply to your build.
+
Alternatively, you could run `gradle help --warning-mode=all` to see the deprecations in the console, though it may not report as much detailed information.
. Update your plugins.
+
Some plugins will break with this new version of Gradle, for example because they use internal APIs that have been removed or changed. The previous step will help you identify potential problems by issuing deprecation warnings when a plugin does try to use a deprecated part of the API.
+
. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_9.0]]
== Upgrading from 8.x and earlier

=== Warnings that are now errors

[[deprecated_configurations_should_not_be_used]]
==== Usage of deprecated configurations is now an error

> Usage of the following `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` are now errors:

Author: I just wanted to add something so that the links in the deprecation messages would go somewhere, and didn't put too much effort into these yet.  I'll come back and revise them after we finalize these change.  I thought the new upgrading guide page wouldn't be visible unless it is linked?

> Usage of the following `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` are now errors:

Reviewer 2: No idea.  I was more curious whether this was something we're doing now as a practice or if this was a one-off.  If the plan is to circle back on this, I don't think there's anything that needs to be resolved now.

> Usage of the following `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` are now errors:

Reviewer: Do we plan to update this now that we're closer to merging?

- `LEGACY`
- `DEPRECATED_CONSUMABLE`
- `DEPRECATED_RESOLVABLE`

=== Removed APIs

=== Potential breaking changes

=== Deprecations

[[custom_configuration_roles]]

> The pre-defined roles in `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` should be used instead of custom roles wherever possible.

Reviewer 3: ```suggestion
Custom roles have been deprecated. Use pre-defined roles from `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` instead.
```

Suggestion: lead with the deprecation, follow with the fix.

[[reserved_configuration_names]]

> The names of standard configuration created for sourcesets by the `java-base` should be considered reserved - do not explicitly create them.

Reviewer 3: ```suggestion
`java-base` plugin sourceset names are reserved. Do not explicitly create custom sourcesets with the same names.
```
I'm not really sure what this sentence is trying to say. Here's my attempt to rewrite it in a clearer way -- hopefully you can revise this to better capture the meaning.
Suggestion: can we add a list of the reserved names, or (better yet) link somewhere that documents the names?

Author: I added an explicit list of all the patterns to try and make it clearer.

[[configurations_allowed_usage]]

> The usage of configurations is meant to be fixed at creation.  Mutating the allowed usage on a configuration is deprecated.

Reviewer 3: I'm also unclear on what we "mutating the allowed usage" means here. Can we link to a resource that explains this better, or maybe provide an example? I would also suggest equivocating with wording like "is meant to be" -- either users should do something, or they shouldn't. We should make the documentation clear.

//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

[[upgrading_version_8]]
= Upgrading your build from Gradle 8.x to the latest

This chapter provides the information you need to migrate your Gradle 8.x builds to the latest Gradle release. For migrating from Gradle 4.x, 5.x, 6.x, or 7.x, see the <<upgrading_version_7.adoc#upgrading_version_7, older migration guide>> first.

We recommend the following steps for all users:

. Try running `gradle help --scan` and view the https://gradle.com/enterprise/releases/2018.4/#identify-usages-of-deprecated-gradle-functionality[deprecations view] of the generated build scan.
+
image::deprecations.png[Deprecations View of a Gradle Build Scan]
+

> This is so that you can see any deprecation warnings that apply to your build.

Reviewer 4: ```suggestion
This is so you can see any deprecation warnings that apply to your build.
```

+
. Run `gradle wrapper --gradle-version {gradleVersion}` to update the project to {gradleVersion}.
. Try to run the project and debug any errors using the <<troubleshooting.adoc#troubleshooting, Troubleshooting Guide>>.

[[changes_9.0]]
== Upgrading from 8.x and earlier

=== Warnings that are now errors

[[deprecated_configurations_should_not_be_used]]
==== Usage of deprecated configurations is now an error

Usage of the following `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` are now errors:

- `LEGACY`
- `DEPRECATED_CONSUMABLE`
- `DEPRECATED_RESOLVABLE`

=== Removed APIs

=== Potential breaking changes

=== Deprecations

[[custom_configuration_roles]]
Custom roles have been deprecated. Use pre-defined roles from `@link:{javadocPath}/javadoc/org/gradle/api/internal/artifacts/configurations/ConfigurationRoles.htm[ConfigurationRoles]` instead.

[[reserved_configuration_names]]

> The `java-base` plugin will create configurations with for every sourceset using these patterns:

Reviewer 4: ```suggestion
The `java-base` plugin will create configurations for every sourceset using these patterns:
```

 - `<SourceSet Name>Implementation`
 - `<SourceSet Name>CompileOnly`
 - `<SourceSet Name>RuntimeOnly`
 - `<SourceSet Name>CompileClasspath`
 - `<SourceSet Name>RuntimeClasspath`
 - `<SourceSet Name>AnnotationProcessor`
 - `<SourceSet Name>RuntimeElements`

> In addition, the `java-library` plugin will create configurations with these patterns:

Reviewer: The java library plugin does not create configuration for source sets other than the main source set. It only creates the `api` and `compileOnlyApi` configurations. 

However, feature variants may create configurations in the format you've listed: https://github.com/gradle/gradle/blob/master/subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmVariantBuilder.java#L170-L176

Author: | Do we plan to update this now that we're closer to merging?

It's no longer needed as a link, so I'll just remove it.

 - `<SourceSet Name>Api`
 - `<SourceSet Name>ApiElements`
 - `<SourceSet Name>CompileOnlyApi`

The names of these standard configurations should be considered reserved - do not explicitly create configurations using them.

See the <<java_plugin.adoc#tab:configurations, The Java Plugin's Dependency Management Reference>> for more information.

[[upgrading_version_8:changes_8.0]]

[[configurations_allowed_usage]]
The usage of configurations should be fixed at creation.  Mutating the allowed usage on a configuration is deprecated.
This includes calling any of the following `Configuration` methods:

- `setCanBeConsumed(boolean)`
- `setCanBeResolved(boolean)`

> It also includes calling the following method on `ConfigurationInternal`:

Reviewer: Should we be calling out gradle internals in the upgrade guide?

Author: I'll remove it.
